### PR TITLE
Add a 'private' flag to mediafiles.

### LINKED ISSUE #2116: Allow hidden/private mediafiles
In our case, we want to upload mediafiles beforehand, but not allow anonymous users to view/play these (yet). A checkmark 'public' or similar, with specific permissions related to it, would allow this.

### PR DESCRIPTION
**Pull request by lesteenman**
Only users with the 'mediafiles.can_see_private' permission can create
and view these private mediafiles.

Fixes #2116.

### CODE REVIEW

**Comment by lesteenman** [comment]
This is a check that fails in a generated migration file; Is that intended?

File: openslides/mediafiles/views.py [hunk 1]
```diff
@@ -15,6 +15,13 @@ class MediafileViewSet(ModelViewSet):
     access_permissions = MediafileAccessPermissions()
     queryset = Mediafile.objects.all()
 
+    def get_queryset(self):
```
Comment (normanjaeckel) [inline openslides/mediafiles/views.py]: I think you also have to add some logic to the file `access_permissions.py` and use the `check_object_permissions()` method. See agenda access_permissions and views for example: There we have the same case. Some people may see hidden agenda items, others don't. We need this different logic to ensure that only authorized users get this data even via our websocket autoupdate process.

File: openslides/mediafiles/views.py [hunk 2]
```diff
@@ -15,6 +15,13 @@ class MediafileViewSet(ModelViewSet):
...
+        user = self.request.user
+        if user.has_perm('mediafiles.can_see_private'):
+            return Mediafile.objects.all()
```
Comment (normanjaeckel) [inline openslides/mediafiles/views.py]: Please use something like `queryset = super().get_queryset()` so that line 16 is still in effect.

File: openslides/users/signals.py [hunk 3]
```diff
@@ -118,6 +118,7 @@ def create_builtin_groups_and_admin(**kwargs):
         'core.can_use_chat',
         'mediafiles.can_manage',
         'mediafiles.can_see',
+        'mediafiles.can_see_private',
```
Comment (normanjaeckel) [inline openslides/users/signals.py]: This permission should also be added to the staff group, see line 178.

File: openslides/mediafiles/models.py [hunk 4]
```diff
@@ -41,6 +44,7 @@ class Meta:
         default_permissions = ()
         permissions = (
             ('can_see', 'Can see the list of files'),
+            ('can_see_private', 'Can see privately listed files'),
```
Comment (emanuelschuetze) [inline openslides/mediafiles/models.py]: Please add the permission string to users/site.js to make the string translateable:
https://github.com/OpenSlides/OpenSlides/blob/master/openslides/users/static/js/users/site.js#L986

**Comment by lesteenman** [comment]
Thank you all for the feedback. After using this feature in Practice at a large conference 2 weeks back, I've found some more issues myself. I will be looking into them, but not within a month due to deadlines for finishing my study. 

If someone else feels like picking it up as well, feel welcome, although I fully intend to do so after finishing my studies.

**Comment by lesteenman** [comment]
Hoping you are still watching this: I've picked up this issue again, but I'm running into some issues as I don't fully grasp how everything fits together.

The main issue at the moment for this merge request: Creating a new mediafile does not pick up on the given private setting. Editing it later and changing the 'private' field does work, however. Any thoughts? From what I could see from the `create` function, it should just work as-is.

**Comment by normanjaeckel** [comment]
May be you have to add some logic to the MediafileForm factory (site.js line 298). The uploadFile method is called inside the MediafileCreateCtrl.

**Comment by lesteenman** [comment]
It seems like it fails on installing versions of requirements...

**Comment by emanuelschuetze** [comment]
@lesteenman Please rebase on top of master. I fixed the travis error caused by outdated requirements (see #2232).

For your problem with creating new mediafile: Maybe you run into this problem #2223?

**Comment by emanuelschuetze** [comment]
@lesteenman please ping me here  if your pull request is ready for review.

**Comment by lesteenman** [comment]
I just realize I completely forgot to ping you. I think it is ready for review.

File: openslides/mediafiles/static/templates/mediafiles/mediafile-list.html [hunk 5]
```diff
@@ -177,6 +177,7 @@ <h1 translate>Files</h1>
           <strong><a ng-href="{{ mediafile.mediafileUrl }}" target="_blank">{{ mediafile.title_or_filename }}</a></strong>
           <br><small><i class="fa fa-file"></i> {{ mediafile.filetype }}</small>
           <br><small><i class="fa fa-database"></i> {{ mediafile.filesize }}</small>
+          <span ng-if="mediafile.private"><br><small><i class="fa fa-lock"></i> Private</small></span>
```
Comment (emanuelschuetze) [inline openslides/mediafiles/static/templates/mediafiles/mediafile-list.html]: Please mark string for translation: `<translate>Private</translate>`

File: openslides/mediafiles/models.py (hunk 4 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/models.py]: Suggestion for shorter translation string: `Can see private files`

File: openslides/users/static/js/users/site.js [hunk 6]
```diff
@@ -989,6 +989,7 @@ angular.module('OpenSlidesApp.users.site', ['OpenSlidesApp.users'])
         gettext('Can see the list of files');
         gettext('Can upload files');
         gettext('Can manage files');
+        gettext('Can see privately listed files');
```
Comment (emanuelschuetze) [inline openslides/users/static/js/users/site.js]: Suggestion for shorter translation string: `Can see private files`

**Comment by emanuelschuetze** [comment]
@lesteenman Some generally remarks:
* The new permission does not work for the list view: A user without this new permission should NOT be able to see private files! But all users can see private marked files... (e.g. guests)
* Please add `(Fixes #2116)` in commit message to closed the associated issue automatically after merge.

**Comment by emanuelschuetze** [comment]
I retested: 
Now the new permission works well after reload: A guest can not see private files. Good.
But one use case does not work very good:
* upload a file (without private flag) -> a guest user can see the file
* now the admin edit the file and switch it to 'private' -> the guest user list view gets a "private" string, but the table row is still visible. Expected: row should be removed/invisible.

And please rebase your branch on top of master (there is a fix in master for handling static media files).
Pleese ping me here if I can retest your PR. Thanks!

**Comment by lesteenman** [comment]
> now the admin edit the file and switch it to 'private' -> the guest user list view gets a "private" string, but the table row is still visible. Expected: row should be removed/invisible.

I was not aware something like this was (easily?) possible. I'll check it out!

**Comment by emanuelschuetze** [comment]
You need autoupdate - like in all other list views. E.g. if you delete a motion - in an other list view the motion is gone. Same should be work with files which set to private. Aim: user should not reload manually the list view page.

**Comment by lesteenman** [comment]
Yes, working on it now. I assume that it should pretend the file is deleted when it is changed to private. If that's not possible, deleting it from the list of mediafiles locally might work... But that seems iffy.

**Comment by emanuelschuetze** [comment]
Just try to use a private filter in ng-repeat and something with os-perms="mediafiles.can_see_private", as e.g.:
``
<tr os-perms="!mediafiles.can_see_private" ng-repeat="mediafile in mediafilesFiltered = (mediafiles | ... | filter: {private: false})" ...
``
Maybe that helps...

**Comment by lesteenman** [comment]
That should work as long as the user can't still access the file once it's private; Then it's safe enough to just filter it out clientside. However, I just noticed that you can still access the URL directly when it's private... Trying to fix that first.

**Comment by lesteenman** [comment]
Ok, it seems like this is not that easy to implement at the moment, as mediafiles are served directly. I'll have to figure out how to check permissions on individual files. For now, I will filter it from the list, and send in a new PR.

**Comment by lesteenman** [comment]
> Just try to use a private filter in ng-repeat and something with os-perms="mediafiles.can_see_private", as e.g.:

Do I understand correctly that this would require duplication of everything within that TR, as you'd need one for those who do have the permission, and one for those who don't?

**Comment by lesteenman** [comment]
@emanuelschuetze, please check it again. I think this implementation is a good improvement on the current situation already, although being able to deny direct access to private mediafiles would be a future improvement.

File: openslides/mediafiles/views.py (hunk 1 above)
Comment (normanjaeckel) [inline openslides/mediafiles/views.py]: Ping.

Comment (lesteenman) [inline openslides/mediafiles/views.py]: Forgot about that bit, added it. However, I'm getting errors when logged in, which aren't caused by my changes (checking out upstream/master causes the same issues). Are these known issues?

What I did was rebase on upstream/master, and run `./manage.py migrate`.

capture

Comment (normanjaeckel) [inline openslides/mediafiles/views.py]: I don't know but it loosk like that you do not have enough permissions. We changed the permissions system some weeks ago. A suitable migration file is still missing.

Try to remove your personal data directory and then run python manage.py start to create a new settings file and database file.

File: openslides/mediafiles/models.py (hunk 4 above)
Comment (emanuelschuetze) [inline openslides/mediafiles/models.py]: This renaming to "Can see private files" is still open.

File: openslides/mediafiles/static/templates/mediafiles/mediafile-list.html [hunk 7]
```diff
@@ -177,6 +177,7 @@ <h1 translate>Files</h1>
           <strong><a ng-href="{{ mediafile.mediafileUrl }}" target="_blank">{{ mediafile.title_or_filename }}</a></strong>
           <br><small><i class="fa fa-file"></i> {{ mediafile.filetype }}</small>
           <br><small><i class="fa fa-database"></i> {{ mediafile.filesize }}</small>
+          <span ng-if="mediafile.private"><br><small><i class="fa fa-lock"></i><translate>Private</translate></small></span>
```
Comment (emanuelschuetze) [inline openslides/mediafiles/static/templates/mediafiles/mediafile-list.html]: Add a whitespace between icon and `<translate>` tag.

**Comment by emanuelschuetze** [comment]
I retested. Looks good. Thanks! I found a minor bug only. How to reproduce it: 
* Upload a new (public) file.
* Look with a guest user (without permission to see private files) to mediafile list view (see "1/1 files").
* Admin changed file to private.
* Guest user does not see private file in list but "0/1 files" is still shown. After reload it shows correctly "0/0 files".
Would be great if you can fix this.

If we cannot prevent the direct access to private files we should at least add a "important note" as help text to the private checkbox in mediafile form (see in users or motion form how to do it). Otherwise the admin could get a unpleasant surprise if he found out that files are not protected for direct access - especially if OpenSlides is used online.

@normanjaeckel  Please review again.

File: openslides/mediafiles/access_permissions.py [hunk 8]
```diff
@@ -18,3 +18,16 @@ def get_serializer_class(self, user=None):
         from .serializers import MediafileSerializer
 
         return MediafileSerializer
+
+    # def get_restricted_data(self, full_data, user):
```
Comment (normanjaeckel) [inline openslides/mediafiles/access_permissions.py]: Do you want to use this code? How do you ensure, that only authorized users get private file data via autoupdate. Please have a look at #2357. This is under heavy development. :-(

File: openslides/mediafiles/static/js/mediafiles/site.js [hunk 9]
```diff
@@ -331,12 +331,27 @@ angular.module('OpenSlidesApp.mediafiles.site', ['ngFileUpload', 'OpenSlidesApp.
...
+
+.filter('privateFilter', [
+    '$filter',
+    'operator',
+    function ($filter, operator) {
+        return function (array) {
+            if (operator.hasPerms('mediafiles.can_see_private')) {
+                return array;
+            }
+            return Array.prototype.filter.call(array, function (item) {
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Just for me to learn something: Why not just `array.filter(function (item) {...})`?

File: openslides/mediafiles/static/templates/mediafiles/mediafile-list.html [hunk 10]
```diff
@@ -155,8 +155,8 @@ <h1 translate>Files</h1>
           </i>
       </tr>
     <tbody>
-      <tr ng-repeat="mediafile in mediafilesFiltered = (mediafiles | osFilter: filter.search : getFilterString |
-          filter: {filetype: filter.showPDFs} | orderBy: sortColumn:reverse)"
+      <tr ng-repeat="mediafile in mediafilesFiltered = (mediafiles | privateFilter | osFilter: filter.search : getFilterString |
+	  filter: {filetype: filter.showPDFs} | orderBy: sortColumn:reverse )"
```
Comment (normanjaeckel) [inline openslides/mediafiles/static/templates/mediafiles/mediafile-list.html]: Unnecessary whitespace and the end.

Why did you change indention?

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 9 above)
Comment (lesteenman) [inline openslides/mediafiles/static/js/mediafiles/site.js]: I think that should actually work in this case. I believe I took this approach from somewhere else in the codebase or online, and is usually what you want to use if you have array-likes rather than actual arrays. So if the files are always an array, that approach is possible.

File: openslides/mediafiles/static/templates/mediafiles/mediafile-list.html (hunk 10 above)
Comment (lesteenman) [inline openslides/mediafiles/static/templates/mediafiles/mediafile-list.html]: It appears I finally forgot to change Vim's whitespace settings before pushing :/

File: openslides/mediafiles/access_permissions.py (hunk 8 above)
Comment (normanjaeckel) [inline openslides/mediafiles/access_permissions.py]: See #2394 .

File: openslides/mediafiles/static/js/mediafiles/site.js (hunk 9 above)
Comment (normanjaeckel) [inline openslides/mediafiles/static/js/mediafiles/site.js]: Ok.

**Comment by normanjaeckel** [comment]
@lesteenman Do you need any help here? We want to refactor many files in about two weeks.

**Comment by lesteenman** [comment]
I'll get to it later today. Am I right that the remaining open issue is:

- The `get_restricted_data` function to prevent others from receiving private files (using something from #2394)

**Comment by lesteenman** [comment]
Not receiving private works, removing files from the list when turned private does not yet...

**Comment by lesteenman** [comment]
@normanjaeckel I just noticed that the same issue occurs with Agenda items that become hidden. Is it safe to assume this is a bug in general?

**Comment by lesteenman** [comment]
Assuming I can, and that these were the remaining issues: Should I merge all commits in this branch myself, or will that be done when it's merged?

File: openslides/mediafiles/models.py (hunk 4 above)
Comment (normanjaeckel) [inline openslides/mediafiles/models.py]: Ping.

File: openslides/mediafiles/static/templates/mediafiles/mediafile-list.html (hunk 7 above)
Comment (normanjaeckel) [inline openslides/mediafiles/static/templates/mediafiles/mediafile-list.html]: Ping

**Comment by normanjaeckel** [comment]
Ok. This is a bug in general. We have to think about it later.

Please also change the permission string in the migrations file. Then rebase all commits to one and rebase it also on top of master.

Thanks. :-)

**Comment by lesteenman** [comment]
Done!

**Comment by emanuelschuetze** [comment]
One migration problem:
I can't see the new checkbox "private" if I migrate my database (created with current master). I cleared my full cache and I have already the right permissions...
If I start with a fresh db in this branch all works well (private checkbox is visible).

@lesteenman: Can you verify this? Please test the migration path. We want to make sure that users with an old database can use the new private feature.

**Comment by lesteenman** [comment]
@emanuelschuetze what is the recommended way to clean an active working directory? I seem to get issues when rebasing on master every time with random issues. I run the server with either `./manage.py start` or `./manage.py runserver`. Just deleting `personal_data` didn't do the trick.

**Comment by emanuelschuetze** [comment]
Just remove the `personal_data` directory and restart openslides with `python manage.py start`.

An example how I reproduced  the migration problem:
```
rm -r personal data
git checkout master
npm install
python manage.py start
# add some mediafiles
# stop server
git checkout <your-privateMediafiles-branch>
npm install
python manage.py start  # migrates database
# reload browser, clear cache
# edit a mediafile or create a new mediafile -> no 'private' field available in form
```

Hope this helps you.

**Comment by lesteenman** [comment]
Thank you, that should help.

**Comment by lesteenman** [comment]
Right, what is going wrong is that the default permissions are not migrated, so even admin users can not see private mediafiles (and thus the flag) by default. I'll check to see how this is set up properly in migration.

**Comment by lesteenman** [comment]
@emanuelschuetze I'm trying to add the permission to the Staff group in [migration file 3](https://github.com/lesteenman/OpenSlides/blob/privateMediafiles/openslides/mediafiles/migrations/0003_auto_20160514_1347.py). However, it seems the permission with codename `can_see_private` does not yet exist here.

What is the recommended way to add a default permission in a migration? [The groups migration](https://github.com/lesteenman/OpenSlides/blob/privateMediafiles/openslides/users/migrations/0004_groups.py) has been the only migration file I've found so far which does something with migrations. It seems that in this case, there's an issue where the new permissions are not yet created when the migration runs, however...

**Comment by normanjaeckel** [comment]
At the moment, it is not possible to add new permissions to groups during migration because Django creates the permission just **after** the migration process. Sorry.

**Comment by lesteenman** [comment]
@normanjaeckel that explains it... So then what can be done? The permission needs to be added to staff somewhere.

**Comment by normanjaeckel** [comment]
Add permission to staff in users/signals.py would do the thing for people who start with a fresh db. All others have to read a nice CHANGELOG message (add this!) and use the interface (users -> groups -> nice interface to control the permissions).

**Comment by emanuelschuetze** [comment]
This is also ok for me.

**Comment by lesteenman** [comment]
Ok. In that case, I think the current branch can be merged.

**Comment by normanjaeckel** [comment]
@ostcar This pull request implies small changes in your cache branch. You have to check the mediafile app and delete the get_queryset() method in the viewset.
